docs/design/yaml-file.rst

```diff
@@ -0,0 +1,158 @@
+YAML Configuration File
+=======================
+
+Backgroud
+---------
+
+The current YAML configuration file is in beta state.
+There are many options and features that it doesn't support yet.
+This document will serve as a design document for discuss how to implement the missing features.
+
+Scope
+-----
+
+- Finish the spec to include all the missing options
+- Have consistency around the spec
+- Proper documentation for the end user
+- Allow to specify the spec's version used on the YAML file
+- Show the YAML file on the build process
+- Show/suggest a YAML file at the project creation if it hasn't one
+- Have one source of truth for global configurations 
+
+RTD settings
+------------
+
+No all the RTD settings are applicable to the YAML file,
+others are applicable for each build (or version),
+and others for the global project. 
+
+Not applicable settings
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Those settings can't be on the YAML file because:
+may depend for the initial project setup,
+are planned to be removed,
+security and privacy reasons.
+
+- Project Name
+- Repo URL
+- Repo type
+- Privacy level (this feature is planned to be removed [#privacy-level]_)
+- Project description (this feature is planned to be removed [#project-description]_)
+- Single version (*)
+- Default branch
+- Default version (*)
+- Domains (*)
+- Active versions (*)
+- Translations
+- Subprojects
+- Integrations
+- Notifications
+
+.. note::
+   The items marked with ``(*)`` can be considered to be global settings.
+   But aren't too relevant right now or aren't related to the builds.
+
+Global settings
+~~~~~~~~~~~~~~~
+
+Those settings will be read from the YAML file on the ``default branch``.
```

Comment:
I think this is the part that I'm the most unsure of. It feels quite magical to have the "default branch" be the place to set global project settings -- but I don't have a better option except to just keep those settings in the database, that I can think of.

It's really tricky to think about the best way to define redirects, for example. I could see reasons for them to be per-branch, but also wanting them to expand across all existing branches.

Reply:
Also, there aren't too many global configurations (relevant ones). Analytics code and redirects are the most relevant here in my opinion.

Reply:
I like the idea of reading the global settings from the default branch. This is the solution I was proposing in another discussion. I think we do need a place different that the DB to keep track/history for these settings and the best place I can think of is the default branch.

Also, I think we shouldn't put here any setting related to the build process. I'd like to be able to build very old versions of the documentation using the specific branch/tag without any failure. For this case, these global settings shouldn't affect the build process.

I'd say that global options that need to be overriden in a specific version for any reason (I don't have one, but Eric mentioned this case) could just have the same section of the default branch and override the particular setting: we can merge the global file with the version specific one.

Reply:
Another question here is: are we planning to remove all the data from the DB or we want to use the YAML from the default branch to _populate_ our DB each time we receive a webhook?

On building process of `stable` we need,

* check out default branch and pull
* load the YAML in memory
* check out `stable`
* show YAML in build commands output
* use that YAML to build

> looks there are some coming and going here

or,

* populate our DB when we receive a webhook on default branch (previously done)
* check out `stable`
* build using DB (as we currently do)

> here we loose the ability to print the YAML

Reply:
But projects without a webhook or failed update from the webhook would leave the default branch outdated and that would be used for build the new project when maybe the user expects at least an error of that happening before the build.

Reply:
Also, didn't understand well, are you proposing to use two files? (one for the global options and other for the regular ones)

Reply:
> Also, didn't understand well, are you proposing to use two files?

No. Just one file with different sections as we are doing now (maybe those sections can be overriden in the per-version yml in case it's needed)

Reply:
> What could be the case for per-version redirects?

Say you rename a file in v2:

* `intro` -> `getting-started`

When a user goes to `/en/v1/intro/` you don't want them to get redirected to `/en/v1/getting-started/`, because it will 404. But on v2, that's exactly what you want.

Reply:
I think we have to keep track of the YAML files, putting them in the DB for each `Version` object isn't a bad idea, but either way we need to keep that state in a persistent way.

I think the bigger question is relying on the `default_branch` to effect other builds. It feels really confusing as a user, and there will be times when we get into a state where the `default_branch` syncing has failed at some point, and there isn't a good way to debug it. I think global state should probably continue to live in the DB itself, since that is 100% obvious and consistent.